hrqb/utils/quickbase.py

```diff
@@ -0,0 +1,144 @@
+        """
+        return self.make_request(requests.get, f"apps/{self.app_id}")
+
+    def get_tables(self) -> pd.DataFrame:
+        """Get all QB Tables as a Dataframe.
+
+        https://developer.quickbase.com/operation/getAppTables
+        """
+        tables = self.make_request(requests.get, f"tables?appId={self.app_id}")
+        return pd.DataFrame(tables)
+
+    def get_table_id(self, name: str) -> str:
+        """Get Table ID from Dataframe of Tables."""
+        tables_df = self.get_tables()
+        return tables_df[tables_df.name == name].iloc[0].id
+
+    def get_table_fields(self, table_id: str) -> pd.DataFrame:
+        """Get all QB Table Fields as a Dataframe.
+
+        https://developer.quickbase.com/operation/getFields
+        """
+        fields = self.make_request(requests.get, f"fields?tableId={table_id}")
+        return pd.DataFrame(fields)
+
+    def get_table_fields_name_to_id(self, table_id: str) -> dict:
+        """Get Field name-to-id map for a Table.
+
+        This method is particularly helpful for upserting data via the QB API, where
+        Field IDs are required instead of Field names.
+        """
+        fields_df = self.get_table_fields(table_id)
+        return {f["label"]: f["id"] for _, f in fields_df.iterrows()}
+
+    def upsert_records(self, upsert_payload: dict) -> dict:
+        """Upsert Records into a Table.
+
+        https://developer.quickbase.com/operation/upsert
+        """
+        return self.make_request(requests.post, "records", json=upsert_payload)
+
+    def prepare_upsert_payload(
+        self,
+        table_id: str,
+        records: list[dict],
+        merge_field: str | None = None,
+    ) -> dict:
+        """Prepare an API payload for upsert.
+
+        https://developer.quickbase.com/operation/upsert
+
+        This method expects a list of dictionaries, one dictionary per record, with a
+        {Field Name:Value} structure.  This method will first retrieve a mapping of
+        Field name-to-ID, then remap the data to a {Field ID:Value} structure.
+
+        Then, return a dictionary payload suitable for the QB upsert API call.
+        """
+        field_map = self.get_table_fields_name_to_id(table_id)
+        mapped_records = []
+        for record in records:
+            mapped_record = {}
+            for field_name, field_value in record.items():
+                if field_id := field_map.get(field_name):
+                    mapped_record[str(field_id)] = {"value": field_value}
+                else:
+                    message = (
+                        f"Field name '{field_name}' not found for Table ID '{table_id}'"
+                    )
+                    raise QBFieldNotFoundError(message)
+            mapped_records.append(mapped_record)
+
+        upsert_payload = {
+            "to": table_id,
+            "data": mapped_records,
+            "fieldsToReturn": list(field_map.values()),
+        }
```

Comment:
In the "Request Body" section of [upsert](https://developer.quickbase.com/operation/upsert) in the QB API guide, it reads:

> Update existing record
For updating an existing record, add a mapping of the table key field ID (defaults to '3') to the record ID, in the data array.

Do you know what is meant by "defaults to '3'"?

Reply:
Does the value for "3" represent a unique record identifier?

Reply:
This is pinpointing the Quickbase pain point.  By default, the primary key for a Quickbase Table _just so happens to be_ Field ID `3`.  But that's not gauranteed.  Said another way, you cannot be sure that Field ID `3` is the primary key for a Table.  

This is why the `QBClient` prepares a Field label-to-ID map when inserting data.  The method `QBClient.prepare_upsert_payload` accepts a parameter `merge_field` which will get translated to the correct Field ID for that Field label.  That allows for updating a row when you, as the developer, know which field should be used to match the row in QB, as it's usually _not_ the primary key `Record ID` (again, which is _usually_ `3` but not always).